# Twitter Data Analysis (Dataset 1)

## Script 1 — Pure Python (No Pandas, No Polars)


In [ ]:
import csv

# Open the file
with open("C:\\Users\\s29sh\\OneDrive\\Documents\\Datasets\\period_03\\2024_tw_posts_president_scored_anon.csv", mode='r', encoding='utf-8') as file:
    reader = csv.reader(file)
    header = next(reader)
    first_row = next(reader)

print("Header:\n", header)
print("\nFirst Row:\n", first_row)


Header:
 ['id', 'url', 'source', 'retweetCount', 'replyCount', 'likeCount', 'quoteCount', 'viewCount', 'createdAt', 'lang', 'bookmarkCount', 'isReply', 'isRetweet', 'isQuote', 'isConversationControlled', 'quoteId', 'inReplyToId', 'month_year', 'illuminating_scored_message', 'election_integrity_Truth_illuminating', 'advocacy_msg_type_illuminating', 'issue_msg_type_illuminating', 'attack_msg_type_illuminating', 'image_msg_type_illuminating', 'cta_msg_type_illuminating', 'engagement_cta_subtype_illuminating', 'fundraising_cta_subtype_illuminating', 'voting_cta_subtype_illuminating', 'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating', 'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating', 'health_topic_illuminating', 'immigration_topic_illuminating', 'lgbtq_issues_topic_illuminating', 'military_topic_illuminating', 'race_and_ethnicity_topic_illuminating', 'safety_topic_illuminating', 'social_and_cultural

In [19]:
##2. Descriptive Statistics for Numeric Columns
import csv
import statistics

csv_file = "2024_tw_posts_president_scored_anon.csv"

metrics = ["retweetCount", "replyCount", "likeCount", "quoteCount", "viewCount"]

metrics_data = {metric: [] for metric in metrics}

with open(csv_file, encoding="utf-8") as f:
    tw_reader = csv.DictReader(f)
    for tweet in tw_reader:
        for metric in metrics:
            val = tweet[metric]
            try:
                metrics_data[metric].append(int(val))
            except (ValueError, TypeError):
                continue  # Skip if not convertible to int

def summary_stats(nums):
    if not nums:
        return None
    return {
        "Count": len(nums),
        "Mean": round(statistics.mean(nums), 2),
        "Min": min(nums),
        "Max": max(nums),
        "Std Dev": round(statistics.stdev(nums), 2) if len(nums) > 1 else 0.0
    }

for metric in metrics:
    print(f"Summary for '{metric}':")
    print(summary_stats(metrics_data[metric]))
    print("-" * 40)


Summary for 'retweetCount':
{'Count': 27304, 'Mean': 1322.06, 'Min': 0, 'Max': 144615, 'Std Dev': 3405.0}
----------------------------------------
Summary for 'replyCount':
{'Count': 27304, 'Mean': 1063.79, 'Min': 0, 'Max': 121270, 'Std Dev': 3174.98}
----------------------------------------
Summary for 'likeCount':
{'Count': 27304, 'Mean': 6913.69, 'Min': 0, 'Max': 915221, 'Std Dev': 21590.31}
----------------------------------------
Summary for 'quoteCount':
{'Count': 27304, 'Mean': 128.08, 'Min': 0, 'Max': 123320, 'Std Dev': 1131.53}
----------------------------------------
Summary for 'viewCount':
{'Count': 27304, 'Mean': 507084.73, 'Min': 5, 'Max': 333502775, 'Std Dev': 3212173.99}
----------------------------------------


In [ ]:
#3. Frequency Counts for Categorical Variables
from collections import Counter

# Categorical columns to analyze
categorical_columns = ["lang", "isReply", "isRetweet", "source"]

# Store frequency counts
category_counts = {col: Counter() for col in categorical_columns}

# Read and count values
with open(csv_file, mode='r', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    for row in reader:
        for col in categorical_columns:
            value = row[col].strip()
            category_counts[col][value] += 1

# Display results
for col in categorical_columns:
    print(f"Column: {col}")
    print(f"Unique values: {len(category_counts[col])}")
    print("Top 5 most common values:")
    for value, count in category_counts[col].most_common(5):
        print(f"  {value}: {count}")
    print("-" * 40)


Column: lang
Unique values: 12
Top 5 most common values:
  en: 27281
  fr: 6
  tl: 4
  es: 3
  da: 3
----------------------------------------
Column: isReply
Unique values: 2
Top 5 most common values:
  False: 23930
  True: 3374
----------------------------------------
Column: isRetweet
Unique values: 1
Top 5 most common values:
  False: 27304
----------------------------------------
Column: source
Unique values: 14
Top 5 most common values:
  Twitter Web App: 14930
  Twitter for iPhone: 8494
  Sprout Social: 2933
  Twitter Media Studio: 499
  Twitter for iPad: 266
----------------------------------------


In [ ]:
#4. Grouped Aggregations by Language

import csv
import statistics

group_by = "lang"
summarize_fields = ["likeCount", "viewCount"]

# Grouping storage
language_metrics = {}

with open(csv_file, encoding="utf-8") as f:
    tweet_reader = csv.DictReader(f)
    for tweet in tweet_reader:
        lang_key = tweet[group_by].strip()
        if lang_key not in language_metrics:
            language_metrics[lang_key] = {field: [] for field in summarize_fields}
        for field in summarize_fields:
            try:
                language_metrics[lang_key][field].append(int(tweet[field]))
            except (ValueError, TypeError):
                continue

# Print group summaries
for lang, field_data in language_metrics.items():
    print(f"Language Group: {lang}")
    for field in summarize_fields:
        vals = field_data[field]
        if vals:
            avg = round(statistics.mean(vals), 2)
            cnt = len(vals)
            print(f"  {field} → Mean: {avg}, Count: {cnt}")
    print("-" * 40)


Language Group: en
  likeCount → Mean: 6913.52, Count: 27281
  viewCount → Mean: 507323.4, Count: 27281
----------------------------------------
Language Group: tl
  likeCount → Mean: 23366.25, Count: 4
  viewCount → Mean: 443447.25, Count: 4
----------------------------------------
Language Group: et
  likeCount → Mean: 47, Count: 1
  viewCount → Mean: 3355, Count: 1
----------------------------------------
Language Group: es
  likeCount → Mean: 1879, Count: 3
  viewCount → Mean: 245237.33, Count: 3
----------------------------------------
Language Group: pl
  likeCount → Mean: 507, Count: 1
  viewCount → Mean: 69531, Count: 1
----------------------------------------
Language Group: fr
  likeCount → Mean: 770, Count: 6
  viewCount → Mean: 41883.67, Count: 6
----------------------------------------
Language Group: in
  likeCount → Mean: 2, Count: 1
  viewCount → Mean: 109, Count: 1
----------------------------------------
Language Group: nl
  likeCount → Mean: 178, Count: 1
  viewCount

## Insights from Output:
The majority of tweets (27,281) were in English (en), making it the most prevalent language in the dataset.

Tagalog (tl) tweets showed a notably high average number of likes (approximately 23,366), but this figure is based on just 4 tweets, suggesting these numbers may be skewed by outliers.

Tweets in other languages, such as Haitian Creole (ht) and Turkish (tr), recorded exceptionally high average view counts, which is likely a result of the small number of tweets in those languages.

## Script 2 — With Pandas

In [5]:
import pandas as pd

# Load the CSV using Pandas
df = pd.read_csv("2024_tw_posts_president_scored_anon.csv")

# Show basic info and preview
print("Dataset shape:", df.shape)
print("\nColumns:\n", df.columns.tolist())
df.head()


Dataset shape: (27304, 47)

Columns:
 ['id', 'url', 'source', 'retweetCount', 'replyCount', 'likeCount', 'quoteCount', 'viewCount', 'createdAt', 'lang', 'bookmarkCount', 'isReply', 'isRetweet', 'isQuote', 'isConversationControlled', 'quoteId', 'inReplyToId', 'month_year', 'illuminating_scored_message', 'election_integrity_Truth_illuminating', 'advocacy_msg_type_illuminating', 'issue_msg_type_illuminating', 'attack_msg_type_illuminating', 'image_msg_type_illuminating', 'cta_msg_type_illuminating', 'engagement_cta_subtype_illuminating', 'fundraising_cta_subtype_illuminating', 'voting_cta_subtype_illuminating', 'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating', 'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating', 'health_topic_illuminating', 'immigration_topic_illuminating', 'lgbtq_issues_topic_illuminating', 'military_topic_illuminating', 'race_and_ethnicity_topic_illuminating', 'safety_topic_illumi

id  \
0  cc46051622b8a9c1b883a3bbf12c640b12ac1cbdc7f48a...   
1  0e3db0c35a290c6df3b737d15882846c108cc80a9b7e5c...   
2  256905919085d2946d5d187abc6cbe81a8abe3384793b3...   
3  a461b32b31e72b222df7fdda0a8e68b0092e31deda33a8...   
4  ca2795ec79d62adc1fff06c4d3bc9da0bbc899e32c9b21...   

                                                 url              source  \
0  f70a206472e9deaf6e313297c1efb891729ced346a0aeb...  Twitter for iPhone   
1  a1962f54943732a0dc006c33b4b6f5764c7085a1282a59...  Twitter for iPhone   
2  4ddbbdb7f4d8ef62fccf3ed20c993bb665b8620fedb089...  Twitter for iPhone   
3  c7e729c427e714baf06d88a2856a1be07d4ff52e1e6334...  Twitter for iPhone   
4  c589bd751d7e1d275901b184087716d3155a582b164308...     Twitter Web App   

   retweetCount  replyCount  likeCount  quoteCount  viewCount  \
0            10          37         94           2      15610   
1           421        1005       2697          60     158324   
2            39         194        332          12      35535   
3            47         332        427          62     199642   
4            17          46        106           3      17917   

             createdAt lang  ...  military_topic_illuminating  \
0  2023-09-30 14:11:00   en  ...                          0.0   
1  2023-09-29 13:27:24   en  ...                          0.0   
2  2023-09-27 20:31:23   en  ...                          0.0   
3  2023-09-26 01:52:40   en  ...                          0.0   
4  2023-09-21 13:24:13   en  ...                          0.0   

   race_and_ethnicity_topic_illuminating  safety_topic_illuminating  \
0                                    0.0                        0.0   
1                                    0.0                        0.0   
2                                    0.0                        0.0   
3                                    0.0                        0.0   
4                                    0.0                        1.0   

   social_and_cultural_topic_illuminating  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     0.0   

   technology_and_privacy_topic_illuminating  womens_issue_topic_illuminating  \
0                                        0.0                              0.0   
1                                        0.0                              0.0   
2                                        0.0                              0.0   
3                                        0.0                              0.0   
4                                        0.0                              0.0   

   incivility_illuminating scam_illuminating freefair_illuminating  \
0                      1.0               0.0                     0   
1                      1.0               0.0                     0   
2                      1.0               0.0                     0   
3                      0.0               0.0                     0   
4                      1.0               0.0                     0   

   fraud_illuminating  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 47 columns]

In [10]:
import polars as pl

df_pl = pl.read_csv("2024_tw_posts_president_scored_anon.csv")


In [11]:
# Descriptive statistics for numeric columns
df_pl.describe()

shape: (9, 48)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ statistic ┆ id        ┆ url       ┆ source    ┆ … ┆ incivilit ┆ scam_illu ┆ freefair_ ┆ fraud_il │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ y_illumin ┆ minating  ┆ illuminat ┆ luminati │
│ str       ┆ str       ┆ str       ┆ str       ┆   ┆ ating     ┆ ---       ┆ ing       ┆ ng       │
│           ┆           ┆           ┆           ┆   ┆ ---       ┆ f64       ┆ ---       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆           ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ count     ┆ 27304     ┆ 27304     ┆ 27304     ┆ … ┆ 26034.0   ┆ 26034.0   ┆ 27304.0   ┆ 27304.0  │
│ null_coun ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 1270.0    ┆ 1270.0    ┆ 0.0       ┆ 0.0      │
│ t         ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ mean      ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.178574  ┆ 0.012368  ┆ 0.001428  ┆ 0.002747 │
│ std       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.383003  ┆ 0.110526  ┆ 0.037767  ┆ 0.052339 │
│ min       ┆ 0000635d0 ┆ 0000179c6 ┆ Canva     ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│           ┆ c9e7bdf89 ┆ b90798f16 ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ dfc13811d ┆ 7528aaaaf ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ 080…      ┆ 678…      ┆           ┆   ┆           ┆           ┆           ┆          │
│ 25%       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ 50%       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ 75%       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ max       ┆ fffbb471d ┆ ffffd63fa ┆ Twitter   ┆ … ┆ 1.0       ┆ 1.0       ┆ 1.0       ┆ 1.0      │
│           ┆ 8b0bd6d99 ┆ 71574c012 ┆ for       ┆   ┆           ┆           ┆           ┆          │
│           ┆ 0b4f9f222 ┆ 7b90e12fd ┆ iPhone    ┆   ┆           ┆           ┆           ┆          │
│           ┆ 83b…      ┆ ba3…      ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [12]:
# Categorical columns to summarize
categorical_cols = ["lang", "isReply", "isRetweet", "source"]

for col in categorical_cols:
    print(f"\nColumn: {col}")
    print("Unique values:", df[col].nunique())
    print("Top 5 most frequent values:")
    print(df[col].value_counts().head(5))
    print("-" * 40)


Column: lang
Unique values: 12
Top 5 most frequent values:
lang
en    27281
fr        6
tl        4
es        3
da        3
Name: count, dtype: int64
----------------------------------------

Column: isReply
Unique values: 2
Top 5 most frequent values:
isReply
False    23930
True      3374
Name: count, dtype: int64
----------------------------------------

Column: isRetweet
Unique values: 1
Top 5 most frequent values:
isRetweet
False    27304
Name: count, dtype: int64
----------------------------------------

Column: source
Unique values: 14
Top 5 most frequent values:
source
Twitter Web App         14930
Twitter for iPhone       8494
Sprout Social            2933
Twitter Media Studio      499
Twitter for iPad          266
Name: count, dtype: int64
----------------------------------------


In [13]:
# Grouped mean likeCount and viewCount by lang
grouped_stats = df.groupby("lang")[["likeCount", "viewCount"]].agg(["mean", "count"]).round(2)
grouped_stats

likeCount          viewCount       
          mean  count        mean  count
lang                                    
da      597.00      3    33529.67      3
en     6913.52  27281   507323.40  27281
es     1879.00      3   245237.33      3
et       47.00      1     3355.00      1
fr      770.00      6    41883.67      6
ht    40516.00      1  1600170.00      1
in        2.00      1      109.00      1
nl      178.00      1    29081.00      1
pl      507.00      1    69531.00      1
pt     2537.00      1   153972.00      1
tl    23366.25      4   443447.25      4
tr    14433.00      1   434191.00      1

## Script 3 — With Polars

In [14]:
# Load the CSV into a Polars DataFrame
df_pl = pl.read_csv("2024_tw_posts_president_scored_anon.csv")

# Show shape and preview
print("Shape:", df_pl.shape)
df_pl.head()

Shape: (27304, 47)


shape: (5, 47)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ id        ┆ url       ┆ source    ┆ retweetCo ┆ … ┆ incivilit ┆ scam_illu ┆ freefair_ ┆ fraud_il │
│ ---       ┆ ---       ┆ ---       ┆ unt       ┆   ┆ y_illumin ┆ minating  ┆ illuminat ┆ luminati │
│ str       ┆ str       ┆ str       ┆ ---       ┆   ┆ ating     ┆ ---       ┆ ing       ┆ ng       │
│           ┆           ┆           ┆ i64       ┆   ┆ ---       ┆ f64       ┆ ---       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆           ┆ i64       ┆ i64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ cc4605162 ┆ f70a20647 ┆ Twitter   ┆ 10        ┆ … ┆ 1.0       ┆ 0.0       ┆ 0         ┆ 0        │
│ 2b8a9c1b8 ┆ 2e9deaf6e ┆ for       ┆           ┆   ┆           ┆           ┆           ┆          │
│ 83a3bbf12 ┆ 313297c1e ┆ iPhone    ┆           ┆   ┆           ┆           ┆           ┆          │
│ c64…      ┆ fb8…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 0e3db0c35 ┆ a1962f549 ┆ Twitter   ┆ 421       ┆ … ┆ 1.0       ┆ 0.0       ┆ 0         ┆ 0        │
│ a290c6df3 ┆ 43732a0dc ┆ for       ┆           ┆   ┆           ┆           ┆           ┆          │
│ b737d1588 ┆ 006c33b4b ┆ iPhone    ┆           ┆   ┆           ┆           ┆           ┆          │
│ 284…      ┆ 6f5…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 256905919 ┆ 4ddbbdb7f ┆ Twitter   ┆ 39        ┆ … ┆ 1.0       ┆ 0.0       ┆ 0         ┆ 0        │
│ 085d2946d ┆ 4d8ef62fc ┆ for       ┆           ┆   ┆           ┆           ┆           ┆          │
│ 5d187abc6 ┆ cf3ed20c9 ┆ iPhone    ┆           ┆   ┆           ┆           ┆           ┆          │
│ cbe…      ┆ 93b…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ a461b32b3 ┆ c7e729c42 ┆ Twitter   ┆ 47        ┆ … ┆ 0.0       ┆ 0.0       ┆ 0         ┆ 0        │
│ 1e72b222d ┆ 7e714baf0 ┆ for       ┆           ┆   ┆           ┆           ┆           ┆          │
│ f7fdda0a8 ┆ 6d88a2856 ┆ iPhone    ┆           ┆   ┆           ┆           ┆           ┆          │
│ e68…      ┆ a1b…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ ca2795ec7 ┆ c589bd751 ┆ Twitter   ┆ 17        ┆ … ┆ 1.0       ┆ 0.0       ┆ 0         ┆ 0        │
│ 9d62adc1f ┆ d7e1d2759 ┆ Web App   ┆           ┆   ┆           ┆           ┆           ┆          │
│ ff06c4d3b ┆ 01b184087 ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ c9d…      ┆ 716…      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [15]:
# Descriptive statistics for numeric columns
df_pl.describe()

shape: (9, 48)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ statistic ┆ id        ┆ url       ┆ source    ┆ … ┆ incivilit ┆ scam_illu ┆ freefair_ ┆ fraud_il │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ y_illumin ┆ minating  ┆ illuminat ┆ luminati │
│ str       ┆ str       ┆ str       ┆ str       ┆   ┆ ating     ┆ ---       ┆ ing       ┆ ng       │
│           ┆           ┆           ┆           ┆   ┆ ---       ┆ f64       ┆ ---       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆           ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ count     ┆ 27304     ┆ 27304     ┆ 27304     ┆ … ┆ 26034.0   ┆ 26034.0   ┆ 27304.0   ┆ 27304.0  │
│ null_coun ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 1270.0    ┆ 1270.0    ┆ 0.0       ┆ 0.0      │
│ t         ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ mean      ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.178574  ┆ 0.012368  ┆ 0.001428  ┆ 0.002747 │
│ std       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.383003  ┆ 0.110526  ┆ 0.037767  ┆ 0.052339 │
│ min       ┆ 0000635d0 ┆ 0000179c6 ┆ Canva     ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│           ┆ c9e7bdf89 ┆ b90798f16 ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ dfc13811d ┆ 7528aaaaf ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ 080…      ┆ 678…      ┆           ┆   ┆           ┆           ┆           ┆          │
│ 25%       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ 50%       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ 75%       ┆ null      ┆ null      ┆ null      ┆ … ┆ 0.0       ┆ 0.0       ┆ 0.0       ┆ 0.0      │
│ max       ┆ fffbb471d ┆ ffffd63fa ┆ Twitter   ┆ … ┆ 1.0       ┆ 1.0       ┆ 1.0       ┆ 1.0      │
│           ┆ 8b0bd6d99 ┆ 71574c012 ┆ for       ┆   ┆           ┆           ┆           ┆          │
│           ┆ 0b4f9f222 ┆ 7b90e12fd ┆ iPhone    ┆   ┆           ┆           ┆           ┆          │
│           ┆ 83b…      ┆ ba3…      ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

In [16]:
# Categorical columns to inspect
cat_cols = ["lang", "source", "isReply", "isRetweet"]

for col in cat_cols:
    print(f"\nColumn: {col}")
    value_counts = (
        df_pl
        .group_by(col) 
        .agg(pl.count())
        .sort("count", descending=True)
        .head(5)
    )
    print(value_counts)
    print("-" * 40)



Column: lang
shape: (5, 2)
┌──────┬───────┐
│ lang ┆ count │
│ ---  ┆ ---   │
│ str  ┆ u32   │
╞══════╪═══════╡
│ en   ┆ 27281 │
│ fr   ┆ 6     │
│ tl   ┆ 4     │
│ es   ┆ 3     │
│ da   ┆ 3     │
└──────┴───────┘
----------------------------------------

Column: source
shape: (5, 2)
┌──────────────────────┬───────┐
│ source               ┆ count │
│ ---                  ┆ ---   │
│ str                  ┆ u32   │
╞══════════════════════╪═══════╡
│ Twitter Web App      ┆ 14930 │
│ Twitter for iPhone   ┆ 8494  │
│ Sprout Social        ┆ 2933  │
│ Twitter Media Studio ┆ 499   │
│ Twitter for iPad     ┆ 266   │
└──────────────────────┴───────┘
----------------------------------------

Column: isReply
shape: (2, 2)
┌─────────┬───────┐
│ isReply ┆ count │
│ ---     ┆ ---   │
│ bool    ┆ u32   │
╞═════════╪═══════╡
│ false   ┆ 23930 │
│ true    ┆ 3374  │
└─────────┴───────┘
----------------------------------------

Column: isRetweet
shape: (1, 2)
┌───────────┬───────┐
│ isRetweet ┆ count │
│

C:\Users\s29sh\AppData\Local\Temp\ipykernel_20200\3844547438.py:9: DeprecationWarning: `pl.count()` is deprecated. Please use `pl.len()` instead.
(Deprecated in version 0.20.5)
  .agg(pl.count())


In [17]:
# Grouped mean and count of likeCount and viewCount by language
grouped_stats = (
    df_pl
    .group_by("lang")
    .agg([
        pl.col("likeCount").mean().alias("likeCount_mean"),
        pl.col("likeCount").count().alias("likeCount_count"),
        pl.col("viewCount").mean().alias("viewCount_mean"),
        pl.col("viewCount").count().alias("viewCount_count")
    ])
    .sort("likeCount_count", descending=True)
)
print(grouped_stats)

shape: (12, 5)
┌──────┬────────────────┬─────────────────┬────────────────┬─────────────────┐
│ lang ┆ likeCount_mean ┆ likeCount_count ┆ viewCount_mean ┆ viewCount_count │
│ ---  ┆ ---            ┆ ---             ┆ ---            ┆ ---             │
│ str  ┆ f64            ┆ u32             ┆ f64            ┆ u32             │
╞══════╪════════════════╪═════════════════╪════════════════╪═════════════════╡
│ en   ┆ 6913.519886    ┆ 27281           ┆ 507323.401525  ┆ 27281           │
│ fr   ┆ 770.0          ┆ 6               ┆ 41883.666667   ┆ 6               │
│ tl   ┆ 23366.25       ┆ 4               ┆ 443447.25      ┆ 4               │
│ da   ┆ 597.0          ┆ 3               ┆ 33529.666667   ┆ 3               │
│ es   ┆ 1879.0         ┆ 3               ┆ 245237.333333  ┆ 3               │
│ …    ┆ …              ┆ …               ┆ …              ┆ …               │
│ et   ┆ 47.0           ┆ 1               ┆ 3355.0         ┆ 1               │
│ tr   ┆ 14433.0        ┆ 1          

## Insights from Output:
We have obtained the mean and total count for both likeCount and viewCount, grouped by each language (lang).

The results are organized so that languages with the highest number of likeCount entries appear at the top.

This approach aligns precisely with the Pandas-based solution from Script 2.

With this, all requirements for Dataset 1 are completed:

Script 1: Initial exploratory analysis

Script 2: Descriptive summary using Pandas

Script 3: Descriptive summary using Polars